From 78fb15a10c2ef55318c9a551e01be3b2ed91046e Mon Sep 17 00:00:00 2001
From: Michael Catanzaro <mcatanzaro@redhat.com>
Date: Wed, 26 Apr 2023 14:12:56 -0500
Subject: [PATCH 1/3] Revert "connection: a connection is disconnected state
 should not be reusable"

This reverts commit a15107a70b4b2b0b5bda241f037d09af403db96e.
---
 libsoup/soup-connection.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/libsoup/soup-connection.c b/libsoup/soup-connection.c
index f08108f6..4007f282 100644
--- a/libsoup/soup-connection.c
+++ b/libsoup/soup-connection.c
@@ -1373,8 +1373,7 @@ soup_connection_is_reusable (SoupConnection *conn)
 {
         SoupConnectionPrivate *priv = soup_connection_get_instance_private (conn);
 
-        return g_atomic_int_get (&priv->state) != SOUP_CONNECTION_DISCONNECTED &&
-                priv->io_data && soup_client_message_io_is_reusable (priv->io_data);
+        return priv->io_data && soup_client_message_io_is_reusable (priv->io_data);
 }
 
 GThread *
-- 
2.40.0

From e49ecfaeed90d9b07e30f618ebd460395c2e8e55 Mon Sep 17 00:00:00 2001
From: Michael Catanzaro <mcatanzaro@redhat.com>
Date: Wed, 26 Apr 2023 14:13:09 -0500
Subject: [PATCH 2/3] Revert "session: do not finish an already finished item"

This reverts commit b9f4f0367704e4d7f6004645c24887a979014f10.
---
 libsoup/soup-session.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libsoup/soup-session.c b/libsoup/soup-session.c
index 42ea1a87..e9285a97 100644
--- a/libsoup/soup-session.c
+++ b/libsoup/soup-session.c
@@ -1469,7 +1469,7 @@ message_completed (SoupMessage *msg, SoupMessageIOCompletion completion, gpointe
         if (item->state == SOUP_MESSAGE_REQUEUED)
                 item->state = SOUP_MESSAGE_RESTARTING;
 
-	if (item->state != SOUP_MESSAGE_RESTARTING && item->state != SOUP_MESSAGE_FINISHED) {
+	if (item->state != SOUP_MESSAGE_RESTARTING) {
 		item->state = SOUP_MESSAGE_FINISHING;
                 soup_session_process_queue_item (item->session, item, !item->async);
 	}
-- 
2.40.0

From 4409feb66aeed2ebf15ceb9ab55a50f19c16e8af Mon Sep 17 00:00:00 2001
From: Michael Catanzaro <mcatanzaro@redhat.com>
Date: Wed, 26 Apr 2023 14:13:21 -0500
Subject: [PATCH 3/3] Revert "session: handle request cancellation earlier"

This reverts commit 517eb93f33e3d5dcf7ba059608114fc9e36b62af.
---
 libsoup/soup-session.c |  6 +-----
 tests/misc-test.c      | 24 ------------------------
 2 files changed, 1 insertion(+), 29 deletions(-)

diff --git a/libsoup/soup-session.c b/libsoup/soup-session.c
index e9285a97..1020feed 100644
--- a/libsoup/soup-session.c
+++ b/libsoup/soup-session.c
@@ -1740,9 +1740,6 @@ soup_session_process_queue_item (SoupSession          *session,
 		if (item->paused)
 			return;
 
-                if (item->state != SOUP_MESSAGE_FINISHING && g_cancellable_is_cancelled (item->cancellable))
-                        item->state = SOUP_MESSAGE_FINISHING;
-
 		switch (item->state) {
 		case SOUP_MESSAGE_STARTING:
 			if (!soup_session_ensure_item_connection (session, item))
@@ -2910,8 +2907,7 @@ conditional_get_ready_cb (SoupSession               *session,
 	stream = soup_session_send_finish (session, result, &error);
 	if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED)) {
 		soup_cache_cancel_conditional_request (data->cache, data->conditional_msg);
-                if (data->item->state != SOUP_MESSAGE_FINISHED)
-                        cancel_cache_response (data->item);
+		cancel_cache_response (data->item);
 		async_cache_conditional_data_free (data);
 		return;
 	}
diff --git a/tests/misc-test.c b/tests/misc-test.c
index 4c14bb87..045161d8 100644
--- a/tests/misc-test.c
+++ b/tests/misc-test.c
@@ -429,14 +429,6 @@ ea_message_starting (SoupMessage  *msg,
 	g_cancellable_cancel (cancellable);
 }
 
-static void
-ea_message_queued (SoupSession  *session,
-                   SoupMessage  *msg,
-                   GCancellable *cancellable)
-{
-        g_cancellable_cancel (cancellable);
-}
-
 static void
 do_early_abort_test (void)
 {
@@ -496,22 +488,6 @@ do_early_abort_test (void)
 	g_object_unref (cancellable);
 	g_object_unref (msg);
 	soup_test_session_abort_unref (session);
-
-        session = soup_test_session_new (NULL);
-        msg = soup_message_new_from_uri ("GET", base_uri);
-        cancellable = g_cancellable_new ();
-
-        g_signal_connect (session, "request-queued",
-                          G_CALLBACK (ea_message_queued), cancellable);
-        g_assert_null (soup_test_session_async_send (session, msg, cancellable, &error));
-        debug_printf (2, "  Message 4 completed\n");
-
-        g_assert_cmpuint (soup_message_get_connection_id (msg), ==, 0);
-        g_assert_error (error, G_IO_ERROR, G_IO_ERROR_CANCELLED);
-        g_clear_error (&error);
-        g_object_unref (cancellable);
-        g_object_unref (msg);
-        soup_test_session_abort_unref (session);
 }
 
 static void
-- 
2.40.0

